### PR TITLE
DisableIfArgument

### LINKED ISSUE #313: Explore use of InvocationInterceptor to disable (parameterized) tests
In #163, I wrote:

> We just updated to Jupiter 5.5 (#278) to use the new [`InvocationInterceptor`](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/InvocationInterceptor.html) API. _If_ that API gets called for individual parameterized test invocations, it would be a very powerful way to implement disabling parameterized tests. The provided [`Invocation`](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/InvocationInterceptor.Invocation.html) allows skipping tests and the [`ReflectiveInvocationContext`](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/ReflectiveInvocationContext.html) seems to give access to the actual arguments. Someone should look into that and possibly create an issue if the approach seems promising. 😊

Yes, someone should.

### PR DESCRIPTION
**Pull request by Michael1993**
Closes #313 

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [x] There is documentation (Javadoc and site documentation; added or updated)
* [X] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [x] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [x] Only one sentence per line (especially in `.adoc` files)
* [x] Javadoc uses formal style, while sites documentation may use informal style (see #265)

Documentation (new extension)
* [x] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [x] The `package-info.java` contains information about the new extension

Code
* [X] Code adheres to code style, naming conventions etc.
* [X] Successful tests cover all changes
* [X] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks (see #164)
* [X] Tests use [AssertJ](https://joel-costigliola.github.io/assertj/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [ ] A prepared commit message exists
* [x] The list of contributions inside `README.md` mentions the new contribution (real name optional) 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

### CODE REVIEW

File: docs/disable-if-parameter.adoc
Comment (Bukama) [inline docs/disable-if-parameter.adoc:1]: Why (almost) all uppercase? 
edit: I see - we do it everywhere.... /edit

File: docs/disable-if-parameter.adoc
Comment (Bukama) [inline docs/disable-if-parameter.adoc:6]:
```suggestion
Unlike the `@Disabled` API provided in JUnit Jupiter, which disables the test on first encounter of the annotation, `@DisableIfParameter` is validated before each execution of a parameterized test .
```
Aside the wrong space, I think this is better to understand.

File: docs/disable-if-parameter.adoc
Comment (Bukama) [inline docs/disable-if-parameter.adoc:7]:
```suggestion
As a consequence, instead of disabling the entire set of parameterized tests, each test is evaluated and possibly disabled individually.
```

File: docs/disable-if-parameter.adoc
Comment (Bukama) [inline docs/disable-if-parameter.adoc:43]:
```suggestion
The first invocation got disabled, because the first parameter contains "bright".
The second execution got skipped, because the other parameter contains "dust".
```

File: docs/disable-if-parameter.adoc
Comment (Bukama) [inline docs/disable-if-parameter.adoc:45]:
```suggestion
If substrings are not powerful enough, you can also use regular expressions, using the `matches` value:
```

File: docs/disable-if-parameter.adoc
Comment (Bukama) [inline docs/disable-if-parameter.adoc:67]:
```suggestion
You can also combine both options.
When doing so, test is disabled if any parameter value contains a substring _or_ matches an expression:
```

File: docs/disable-if-parameter.adoc
Comment (Bukama) [inline docs/disable-if-parameter.adoc:83]: To align this tests with the others, and because the regex is not so easy to read like the ones above I would suggest to also add an explanation, which tests are executed and which are not.

File: src/main/java/org/junitpioneer/jupiter/params/DisableIfParameterExtension.java
Comment (Bukama) [inline src/main/java/org/junitpioneer/jupiter/params/DisableIfParameterExtension.java:37]: To me the messages is misleading. If the annotation is present, e.g. with `contains`, then for me it is specified, but as an empty set. I would rephrase to something like "specified with at least one element, but both are empty"

File: src/main/java/org/junitpioneer/jupiter/params/DisableIfParameterExtension.java
Comment (Bukama) [inline src/main/java/org/junitpioneer/jupiter/params/DisableIfParameterExtension.java:38]: To other reviewers: I asked @Michael1993 to add those comments, while we was working on the extension on stream, because the two checks look realy the same, but differ in one word in the middle of them, which I didn't mention at the beginning. With the comments I think its more readable.

File: src/test/java/org/junitpioneer/jupiter/params/DisableIfParameterExtensionTests.java
Comment (Bukama) [inline src/test/java/org/junitpioneer/jupiter/params/DisableIfParameterExtensionTests.java:98]: I'm confused why
1) you don't check for the particular exceptions you throw
2) Why sonar claims full coverage (speaking of the part with the three different exceptions)

🤷‍♂️

**Review by Bukama — CHANGES_REQUESTED** [review]
Only minor things (no wonder after the stream ;) )

File: src/main/java/org/junitpioneer/jupiter/params/DisableIfParameter.java
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/params/DisableIfParameter.java:35]: Kudos for using `<em>` instead of `<i>`! 👍

File: src/main/java/org/junitpioneer/jupiter/params/DisableIfParameter.java
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/params/DisableIfParameter.java:46]: Does it make sense to add minimal Javadoc referring to `String#contains` and `String#matches`?

File: docs/docs-nav.yml
Comment (beatngu13) [inline docs/docs-nav.yml:13]: I like that you refer to parameterized tests, maybe adapt the title here as well (i.e. "Disable Parameterized Test Based on Display Name").

File: src/test/java/org/junitpioneer/jupiter/params/DisableIfParameterExtensionTests.java
Comment (Michael1993) [reply src/test/java/org/junitpioneer/jupiter/params/DisableIfParameterExtensionTests.java:98]: Do you mean the `TestAbortedException`?

Comment (Bukama) [reply src/test/java/org/junitpioneer/jupiter/params/DisableIfParameterExtensionTests.java:98]: I mean you have three checks about misconfiguration you do in the extension, but only one test case and I don't get why SONAR does not complain 🤷‍♂️

Comment (Michael1993) [reply src/test/java/org/junitpioneer/jupiter/params/DisableIfParameterExtensionTests.java:98]: It's line coverage, not part coverage. If the line is called, it's considered covered. Example:
```java
/* 100% coverage */
Optional.of(new Object())
.orElseThrow(() -> new RuntimeException("Never thrown!"));

/* < 100% coverage */
Optional.of(new Object())
.orElseThrow(
    () -> new RuntimeException("Never thrown!")
);
```

**Comment by nipafx** [comment]
Re documentation: There's a bit of textual and a lot of topical overlap between the two `@DisableIf...` extensions. Let's merge them into `disable-parameterized-test-....adoc` (with a similar title. _But_ we need to keep `disable-if-display-name.adoc` around, so we don't break links. Let's just empty the file and point towards the new documentation.

**Review by Bukama — APPROVED** [review]
LGTM

**Comment by Michael1993** [comment]
> LGTM (except merge conflict ;) )

Sorry, could you take a look at the issue (#313)? There is an ongoing discussion there about why this PR is incomplete - I'd appreciate your input.

**Comment by nipafx** [comment]
I hope [to have answered](https://github.com/junit-pioneer/junit-pioneer/issues/313#issuecomment-749758736) all your questions. :)

**Comment by Bukama** [comment]
Just re-request my review if your done @Michael1993

**Comment by Michael1993** [comment]
Things still on TODO:
* [ ] Add JavaDoc on annotations
* [ ] Add new test file and gradle instructions to compile it with parameter naming information

File: README.md
Comment (Bukama) [inline README.md:128]: I call this OK, cause the PR is from 2020 :D

File: docs/disable-parameterized-tests.adoc
Comment (Bukama) [inline docs/disable-parameterized-tests.adoc:79]: Correct placeholder

File: docs/disable-parameterized-tests.adoc
Comment (Bukama) [inline docs/disable-parameterized-tests.adoc:132]: I would add (yes add, not move) the information about case-sensetive to the general information about the disableIfParameter too, cause important.

So the general information and here an example for it

File: docs/disable-parameterized-tests.adoc
Comment (Bukama) [inline docs/disable-parameterized-tests.adoc:181]: I personally like to write this `.. starting from 0 (zero)` to show that I don't have a typo.

File: docs/disable-parameterized-tests.adoc
Comment (Bukama) [inline docs/disable-parameterized-tests.adoc:245]: I would add some kind of theader here, because the not is not a note for the particular example above

**Review by Bukama — APPROVED** [review]
I call this an approve as my comments are mostly personal point of view. I leave it up to you to update the docs.

Before merging we have to be aware of the placeholder

File: docs/disable-parameterized-tests.adoc
Comment (Michael1993) [reply docs/disable-parameterized-tests.adoc:246]: I don't know what you mean. Could you clarify?

**Comment by Bukama** [comment]
Can you update the branch with current main to run with the changed java/ JUnit versions please.

File: docs/disable-parameterized-tests.adoc
Comment (Bukama) [reply docs/disable-parameterized-tests.adoc:246]: For me I see some kind of break between line 244 (Just like..) and 246 (NOTE:) where the sentences are not connected by each other but from the formatting it seems like they belong together.

**Comment by nipafx** [comment]
Looks really good, great work! Some small changes aside, I renamed it from `@DisableIfParameter` etc to `@DisableIfArgument` etc because the conditions target arguments (i.e. the values passed for the parameters) and not the parameters themselves.

**Comment by nipafx** [comment]
We still need a commit message. 😬

**Pull request merged**
